### PR TITLE
Add Fabric Geomean Speedup Statistics

### LINKED ISSUE #29040: Improve fabric perf test reporting
Update performance tests to report additional information:
Report the geomean speedup of the test suite. Report the geomean speedup of the following categories:

(each indent indicates getting the geomean of that category as a sub-category of the parent)
- Full test list
  - Per fabric type (e.g. bucket by line vs ring vs mesh vs 2D topology)
   - per message type
   - per message size

### PR DESCRIPTION
**Pull request by jhaiTT**
### Ticket
https://github.com/tenstorrent/tt-metal/issues/29040

### Problem description
The TT-Fabric test infra currently verifies test results against golden results using percent-difference comparisons. While this helps users verify whether test results are consistent, it does not give users any idea of the overall impact of their changes (to fabric, metal, etc). on fabric performance. 

### What's changed
- Added post-comparison statistics generation in `generate_comparison_statistics`, which is called by  `generate_bandwidth_summary` after golden comparisons
- Modified `populate_comparison_result_bandwidth` to calculate per-test speedup compared to golden results, in addition to difference_percent.
- `generate_comparison_statistics` groups the per-test speedups by topology, sub-grouped by `packet_size` and `ntype`.
- Then, the function computes the geometric mean speedup for the following groups:
1. Overall Geomean Speedup 
2. Geomean Speedup per topology
    1. Geomean Speedup per topology, per packet size
    2. Geomean Speedup per topology, per ntype
- Per-test speedups added to `bandwidth_summary_results...diff.csv`, geomean results are written to `stdout`. 
- Added `generate_comparison_statistics_csv`, which writes the results to `comparison_statistics_<arch_name>.csv` in the `generated/fabric` folder, forming a table like the one below:

| topology | packet_size | ntype                        | geomean_speedup |
|:--------:|:-----------:|:----------------------------:|:---------------:|
| Mesh     |        2048 | ALL                          |        0.998264 |
| Mesh     |        4096 | ALL                          |        0.998243 |
| Mesh     | ALL         | NOC_UNICAST_WRITE            |        0.998548 |
| Mesh     | ALL         | NOC_FUSED_UNICAST_ATOMIC_INC |        0.997715 |
| Mesh     | ALL         | NOC_UNICAST_SCATTER_WRITE    |        0.998498 |
| Mesh     | ALL         | ALL                          |        0.998254 |
| Ring     |        2048 | ALL                          |        0.999444 |
| Ring     |        4096 | ALL                          |         1.00364 |
| Ring     | ALL         | NOC_UNICAST_WRITE            |         1.00026 |
| Ring     | ALL         | NOC_FUSED_UNICAST_ATOMIC_INC |         1.00435 |
| Ring     | ALL         | NOC_UNICAST_SCATTER_WRITE    |         1.00002 |
| Ring     | ALL         | ALL                          |         1.00154 |
| Linear   |        2048 | ALL                          |        0.999877 |
| Linear   |        4096 | ALL                          |         1.00004 |
| Linear   | ALL         | NOC_UNICAST_WRITE            |         0.99996 |
| Linear   | ALL         | NOC_FUSED_UNICAST_ATOMIC_INC |         1.00007 |
| Linear   | ALL         | NOC_UNICAST_SCATTER_WRITE    |        0.999844 |
| Linear   | ALL         | ALL                          |        0.999957 |
| Overall  | ALL         | ALL                          |        0.999916 |

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes https://github.com/tenstorrent/tt-metal/actions/runs/18502921513
- [x] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable) https://github.com/tenstorrent/tt-metal/actions/runs/18502930168
- [x] [Blackhole Nightly](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-nightly-tests.yaml) https://github.com/tenstorrent/tt-metal/actions/runs/18502936667
- [x] [Galaxy quick](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-quick.yaml) CI passes (if applicable) https://github.com/tenstorrent/tt-metal/actions/runs/18502942708

### CODE REVIEW

File: tests/tt_metal/tt_metal/perf_microbenchmark/routing/tt_fabric_test_context.cpp [hunk 1]
```diff
@@ -318,3 +318,147 @@ std::string TestContext::generate_failed_test_format_string(const BandwidthResul
...
+        speedups_per_topology_[result_topology.value()].speedups_by_packet_size[result.packet_size].push_back(
+            result_speedup);
+        std::optional<NocSendType> result_ntype = enchantum::cast<NocSendType>(result.ntype);
+        TT_FATAL(result_ntype.has_value(), "Invalid ntype in comparison result: {}", result.ntype);
+        speedups_per_topology_[result_topology.value()].speedups_by_ntype[result_ntype.value()].push_back(
+            result_speedup);
+    }
+}
+
+double TestContext::calculate_geomean_speedup(const std::vector<double>& speedups) {
```
Comment (SeanNijjar) [inline tests/tt_metal/tt_metal/perf_microbenchmark/routing/tt_fabric_test_context.cpp]: Can you please definitel outline this to a separate analysis source file?

File: tests/tt_metal/tt_metal/perf_microbenchmark/routing/tt_fabric_test_context.cpp [hunk 2]
```diff
@@ -318,3 +318,147 @@ std::string TestContext::generate_failed_test_format_string(const BandwidthResul
...
+    comparison_statistics_csv_stream << "topology,packet_size,ntype,geomean_speedup\n";
+    log_info(
+        tt::LogTest, "Initialized comparison statistics CSV file: {}", comparison_statistics_csv_file_path_.string());
+
+    // Write most specific speedups first, then overall geomean speedup
+    for (const auto& [topology, topology_speedups] : speedups_per_topology_) {
+        std::string topology_str = std::string(enchantum::to_string(topology));
+        for (const auto& [packet_size, speedup] : topology_speedups.geomean_speedup_by_packet_size) {
+            comparison_statistics_csv_stream << topology_str << "," << packet_size << "," << "ALL" << "," << speedup
+                                             << "\n";
```
[the remark this answers is not in the document]
Comment (jhaiTT) [reply tests/tt_metal/tt_metal/perf_microbenchmark/routing/tt_fabric_test_context.cpp]: Not implemented, because the string is used as a CSV label, not as a reference.

File: tests/tt_metal/tt_metal/perf_microbenchmark/routing/tt_fabric_test_context.cpp [hunk 3]
```diff
@@ -318,3 +318,147 @@ std::string TestContext::generate_failed_test_format_string(const BandwidthResul
         + tolerance_stream.str();
     return csv_format_string;
 }
+
+void TestContext::organize_speedups_by_topology() {
```
Comment (SeanNijjar) [inline tests/tt_metal/tt_metal/perf_microbenchmark/routing/tt_fabric_test_context.cpp]: Can the bulk of these changes be moved to some new results analysis file? I guess you could build a new type or top level function that takes `comparison_results_` as an arg and does all the work you've put in here. I'm mainly thinking we should try to bloat the test context too much.

File: tests/tt_metal/tt_metal/perf_microbenchmark/routing/tt_fabric_test_context.hpp [hunk 4]
```diff
@@ -153,10 +153,20 @@ struct ComparisonResult {
     double current_bandwidth_GB_s{};
     double golden_bandwidth_GB_s{};
     double difference_percent{};
+    double speedup{};
```
Comment (SeanNijjar) [inline tests/tt_metal/tt_metal/perf_microbenchmark/routing/tt_fabric_test_context.hpp]: I think this should just be a method that is computed based on `golden_bandwidth_GB_s` and `current_bandwidth_GB_s`. (on that note, so should `difference_percent` but that wasn't part of this PR.

File: tests/tt_metal/tt_metal/perf_microbenchmark/routing/tt_fabric_test_context.hpp [hunk 5]
```diff
@@ -153,10 +153,20 @@ struct ComparisonResult {
...
     bool within_tolerance{};
     std::string status;
 };
 
+struct SpeedupsByTopology {
```
Comment (SeanNijjar) [inline tests/tt_metal/tt_metal/perf_microbenchmark/routing/tt_fabric_test_context.hpp]: move to new analysis file.

File: tests/tt_metal/tt_metal/perf_microbenchmark/routing/tt_fabric_test_context.hpp [hunk 6]
```diff
@@ -1437,20 +1457,35 @@ class TestContext {
         }
         // Write diff header
         diff_csv_stream << "test_name,ftype,ntype,topology,num_devices,num_links,packet_size,num_iterations,"
-                           "current_avg_bandwidth_gb_s,golden_avg_bandwidth_gb_s,difference_percent,status\n";
+                           "current_avg_bandwidth_gb_s,golden_avg_bandwidth_gb_s,speedup,difference_percent,status\n";
```
Comment (SeanNijjar) [inline tests/tt_metal/tt_metal/perf_microbenchmark/routing/tt_fabric_test_context.hpp]: I think for this CSV we can just keep speedup out since it's largely redundant with `difference_percent`

File: tests/tt_metal/tt_metal/perf_microbenchmark/routing/tt_fabric_test_context.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (jhaiTT) [reply tests/tt_metal/tt_metal/perf_microbenchmark/routing/tt_fabric_test_context.cpp]: Implemented a fix and error log in commit: 6439428d499964c7ff144c24340e2735745eb807

File: tests/tt_metal/tt_metal/perf_microbenchmark/routing/tt_fabric_test_context.cpp [hunk 7]
```diff
@@ -318,3 +318,147 @@ std::string TestContext::generate_failed_test_format_string(const BandwidthResul
...
+}
+
+double TestContext::calculate_geomean_speedup(const std::vector<double>& speedups) {
+    double geomean_speedup = 1.0;
+    for (const auto& speedup : speedups) {
+        geomean_speedup *= speedup;
+    }
+    geomean_speedup = std::pow(geomean_speedup, 1.0 / speedups.size());
+    return geomean_speedup;
+}
```
[the remark this answers is not in the document]
Comment (jhaiTT) [reply tests/tt_metal/tt_metal/perf_microbenchmark/routing/tt_fabric_test_context.cpp]: Implemented in commit: 6439428d499964c7ff144c24340e2735745eb807

File: tests/tt_metal/tt_metal/perf_microbenchmark/routing/tt_fabric_test_context.hpp (hunk 6 above)
Comment (jhaiTT) [reply tests/tt_metal/tt_metal/perf_microbenchmark/routing/tt_fabric_test_context.hpp]: Ok, removed in commit: d38b7540ba2dc55a66f86252180b6f2e694497ad

File: tests/tt_metal/tt_metal/perf_microbenchmark/routing/tt_fabric_test_context.hpp (hunk 4 above)
Comment (jhaiTT) [reply tests/tt_metal/tt_metal/perf_microbenchmark/routing/tt_fabric_test_context.hpp]: Ok, implemented in commit: d38b7540ba2dc55a66f86252180b6f2e694497ad. Would you like me to do the same with `difference_percent` and `within_tolerance`, or just leave those for a separate PR?

Comment (SeanNijjar) [reply tests/tt_metal/tt_metal/perf_microbenchmark/routing/tt_fabric_test_context.hpp]: up to you. If it's not a blocker, you may as well do it now. If it ends up being invasive, we can do it separately.

File: tests/tt_metal/tt_metal/perf_microbenchmark/routing/tt_fabric_test_context.hpp (hunk 5 above)
Comment (jhaiTT) [reply tests/tt_metal/tt_metal/perf_microbenchmark/routing/tt_fabric_test_context.hpp]: Moved to `tt_fabric_test_results.hpp` in commit: 476568710a526e46096cfb9e1d4af481fffb95db

File: tests/tt_metal/tt_metal/perf_microbenchmark/routing/tt_fabric_test_context.cpp (hunk 3 above)
Comment (jhaiTT) [reply tests/tt_metal/tt_metal/perf_microbenchmark/routing/tt_fabric_test_context.cpp]: Moved all speedup-related operations to `tt_fabric_test_results.hpp` in commit: 476568710a526e46096cfb9e1d4af481fffb95db. This also required moving the `ComparisonResult` structure to the new file, so I moved all the structs (`BandwidthResult`, `BandwidthResultSummary`, `GoldenCSVEntry`, `ComparisonResult`) to the new file, to keep everything in 1 place.

File: tests/tt_metal/tt_metal/perf_microbenchmark/routing/tt_fabric_test_context.cpp (hunk 1 above)
Comment (jhaiTT) [reply tests/tt_metal/tt_metal/perf_microbenchmark/routing/tt_fabric_test_context.cpp]: Moved to `tt_fabric_test_results.cpp` in commit: 476568710a526e46096cfb9e1d4af481fffb95db

File: tests/tt_metal/tt_metal/perf_microbenchmark/routing/tt_fabric_test_context.hpp (hunk 4 above)
Comment (jhaiTT) [reply tests/tt_metal/tt_metal/perf_microbenchmark/routing/tt_fabric_test_context.hpp]: I changed `difference_percent` to be computed as a method as well, but I left `within_tolerance` as an externally-set variable. This is to prevent an edge case, when a golden entry is not found. In that case, the golden BW is set to 0.0. If (in some remote case) the test result's bandwidth is also 0.0 (due to an error, etc), these two values will be shown as within tolerance of one another, and thus the test infra won't show an error condition. Implemented here: 476568710a526e46096cfb9e1d4af481fffb95db

**Review by SeanNijjar — APPROVED** [review]

**Review by nhuang-tt — APPROVED** [review]
